### PR TITLE
feat(otelaws): add sns attribute instrumentation

### PR DESCRIPTION
**Pull request by pragmaticivan**
**Notes:**

* Only the `semconv/v1.27.0` was available in the latest published version of the otel module.
* The current project setup doesn't take into consideration all the various hooks (`m.initializeMiddlewareBefore, m.initializeMiddlewareAfter, m.finalizeMiddlewareAfter, m.deserializeMiddleware`). All attributes are currently being set on `initializeMiddlewareAfter`.
* Ref: https://opentelemetry.io/docs/specs/semconv/attributes-registry/messaging/
* Added changelog

### CODE REVIEW

**Comment by dmathieu** [comment]
This will need a changelog entry.

**Comment by pragmaticivan** [comment]
Thanks @dmathieu !

File: CHANGELOG.md [hunk 1]
```diff
@@ -41,7 +42,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   - Use `baggagecopy.NewLogProcessor` when configuring a Log Provider.
     - `NewLogProcessor` accepts a `Filter` function type that selects which baggage members are added to the log record.
 
-### Changed 
+### Changed
```
Comment (MrAlias) [inline CHANGELOG.md:45]: Please move this change to it's own PR.

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/service/sns"
+	"github.com/aws/smithy-go/middleware"
+
+	"go.opentelemetry.io/otel/attribute"
+	semconv "go.opentelemetry.io/otel/semconv/v1.27.0"
+)
+
+// SNSAttributeSetter sets SNS specific attributes depending on the SNS operation being performed.
```
Comment (MrAlias) [inline instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go:17]:
```suggestion
// SNSAttributeSetter sets SNS specific attributes depending on the SNS operation is being performed.
```

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+func SNSAttributeSetter(ctx context.Context, in middleware.InitializeInput) []attribute.KeyValue {
+	snsAttributes := []attribute.KeyValue{semconv.MessagingSystemKey.String("aws.sns")}
```
Comment (MrAlias) [inline instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go:19]: This is [`aws_sqs` for SQS](https://github.com/open-telemetry/semantic-conventions/blob/0edb9e2b42e37739fe53e10623d93809269dd300/model/messaging/registry.yaml#L167).

```suggestion
	snsAttributes := []attribute.KeyValue{semconv.MessagingSystemKey.String("aws_sns")}
```

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+			semconv.MessagingDestinationName(extractDestinationName(v.TopicArn, v.TargetArn, v.PhoneNumber)),
+			semconv.MessagingOperationTypePublish,
+			semconv.MessagingOperationName("publish_input"),
+		)
+	}
+
+	return snsAttributes
+}
+
+func extractDestinationName(topicArn *string, targetArn *string, phoneNumber *string) string {
```
Comment (MrAlias) [inline instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go:40]: nit

```suggestion
func extractDestinationName(topicArn, targetArn, phoneNumber *string) string {
```

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	if topicArn != nil && *topicArn != "" {
+		return (*topicArn)[strings.LastIndex(*topicArn, ":")+1:]
+	} else if targetArn != nil && *targetArn != "" {
+		return (*targetArn)[strings.LastIndex(*targetArn, ":")+1:]
+	} else if phoneNumber != nil {
+		return *phoneNumber
+	}
+	return ""
```
Comment (MrAlias) [inline instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go:48]: Missing test for `extractDestinationName(nil, nil, nil)`

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes_test.go [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+)
+
+func TestPublishInput(t *testing.T) {
+	input := middleware.InitializeInput{
+		Parameters: &sns.PublishInput{
+			TopicArn: aws.String("arn:aws:sns:us-east-2:444455556666:my-topic"),
+		},
+	}
+
+	attributes := SNSAttributeSetter(context.TODO(), input)
```
Comment (MrAlias) [inline instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes_test.go:26]: Why `context.TODO` and not `context.Background` here?

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes_test.go [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+	input := middleware.InitializeInput{
+		Parameters: &sns.PublishBatchInput{
+			TopicArn:                   aws.String("arn:aws:sns:us-east-2:444455556666:my-topic-batch"),
+			PublishBatchRequestEntries: []types.PublishBatchRequestEntry{},
+		},
+	}
+
+	attributes := SNSAttributeSetter(context.TODO(), input)
+
+	assert.Contains(t, attributes, semconv.MessagingDestinationName("my-topic-batch"))
```
Comment (MrAlias) [inline instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes_test.go:43]: Missing assertion of `MessagingSystemKey`.

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes_test.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+func TestPublishInput(t *testing.T) {
+	input := middleware.InitializeInput{
+		Parameters: &sns.PublishInput{
+			TopicArn: aws.String("arn:aws:sns:us-east-2:444455556666:my-topic"),
+		},
+	}
+
+	attributes := SNSAttributeSetter(context.TODO(), input)
+
+	assert.Contains(t, attributes, semconv.MessagingDestinationName("my-topic"))
```
Comment (MrAlias) [inline instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes_test.go:28]: Missing assertion of `MessagingSystemKey`.

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes_test.go (hunk 6 above)
Comment (pragmaticivan) [reply instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes_test.go:26]: 👋🏽 Happy to change. context.TODO was being used in SQS https://github.com/open-telemetry/opentelemetry-go-contrib/blob/main/instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/sqsattributes_test.go#L37

File: instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+	return snsAttributes
+}
+
+func extractDestinationName(topicArn, targetArn, phoneNumber *string) string {
+	if topicArn != nil && *topicArn != "" {
+		return (*topicArn)[strings.LastIndex(*topicArn, ":")+1:]
+	} else if targetArn != nil && *targetArn != "" {
+		return (*targetArn)[strings.LastIndex(*targetArn, ":")+1:]
+	} else if phoneNumber != nil {
```
Comment (akats7) [inline instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go:45]: Do we want to make phone number something that we can disable. Maybe allow for the use 'SMS' or something along those lines as a default.

Comment (pragmaticivan) [reply instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go:45]: Hi @akats7 this will use Phone Number `only` if you don't provide topicArn or targetArn, but if phone number is not present, it will default to an empty string.

Comment (pragmaticivan) [reply instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go:45]: Btw, used a similar logic to the one in JS https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/plugins/node/opentelemetry-instrumentation-aws-sdk/src/services/sns.ts#L49

Comment (akats7) [reply instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go:45]: Yep I understand, I'm just thinking in cases when users don't want phone numbers as attributes, either for PII or cardinality reasons, the cardinality reason probably applies to topic as well.

Comment (MrAlias) [reply instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go:45]: I'm on board with being conservative to start and not include the phone number for now. If the semconv group decides otherwise we can add it in then, right?

Comment (pragmaticivan) [reply instrumentation/github.com/aws/aws-sdk-go-v2/otelaws/snsattributes.go:45]: That sounds good to me; I will make a change

**Review by MrAlias — APPROVED** [review]

**Comment by pragmaticivan** [comment]
Checking what's going with the tests, something odd there after the merge conflict.

**Comment by pragmaticivan** [comment]
Screenshot 2024-12-08 at 11 47 10 AM
Fixed, no errors at the end.

**Review by dmathieu — APPROVED** [review]

**Comment by MrAlias** [comment]
@akats7 does this look good to you?

**Review by akats7 — APPROVED** [review]

**Comment by akats7** [comment]
> @akats7 does this look good to you?

yep looks good, meant to approve earlier
